framework/include/base/ExecuteMooseObjectStorage.h

```diff
@@ -0,0 +1,243 @@
+{
+}
+
+
+template<typename T>
+const MooseObjectStorage<T> &
+ExecuteMooseObjectStorage<T>::operator[](ExecFlagType exec_flag) const
+{
+  // Use find to avoid accidental insertion
+  typename std::map<ExecFlagType, MooseObjectStorage<T> >::const_iterator iter = _execute_objects.find(exec_flag);
+
+  if (iter == _execute_objects.end())
+    mooseError("Unable to locate the desired execute flag, the global list of execute parameters is likely out-of-date.");
+
+  return iter->second;
+}
+
+template<typename T>
+MooseObjectStorage<T> &
+ExecuteMooseObjectStorage<T>::operator[](ExecFlagType exec_flag)
+{
+  // Use find to avoid accidental insertion
+  typename std::map<ExecFlagType, MooseObjectStorage<T> >::iterator iter = _execute_objects.find(exec_flag);
+
+  if (iter == _execute_objects.end())
+    mooseError("Unable to locate the desired execute flag, the global list of execute parameters is likely out-of-date.");
+
+  return iter->second;
+}
+
+
+template<typename T>
+void
+ExecuteMooseObjectStorage<T>::updateActive(THREAD_ID tid/* = 0 */)
+{
+  // Update all objects active list
+  MooseObjectStorage<T>::updateActive(tid);
+
+  // Update the execute flag lists of objects
+  typename std::map<ExecFlagType, MooseObjectStorage<T> >::iterator iter;
+  for (iter = _execute_objects.begin(); iter != _execute_objects.end(); ++iter)
+    iter->second.updateActive(tid);
+}
+
+
+template<typename T>
+void
+ExecuteMooseObjectStorage<T>::jacobianSetup(THREAD_ID tid/* = 0*/) const
+{
+  checkThreadID(tid);
+  typename std::map<ExecFlagType, MooseObjectStorage<T> >::const_iterator iter = _execute_objects.find(EXEC_NONLINEAR);
+  if (iter != _execute_objects.end())
+    iter->second.jacobianSetup(tid);
+}
+
+
+template<typename T>
+void
+ExecuteMooseObjectStorage<T>::residualSetup(THREAD_ID tid/* = 0*/) const
+{
+  checkThreadID(tid);
+  typename std::map<ExecFlagType, MooseObjectStorage<T> >::const_iterator iter = _execute_objects.find(EXEC_LINEAR);
+  if (iter != _execute_objects.end())
+    iter->second.residualSetup(tid);
+}
+
+
+template<typename T>
+void
+ExecuteMooseObjectStorage<T>::setup(ExecFlagType exec_flag, THREAD_ID tid/* = 0*/) const
+{
+  checkThreadID(tid);
+  switch (exec_flag)
+  {
+  case EXEC_INITIAL:
+    initialSetup(tid);
+    break;
+  case EXEC_TIMESTEP_BEGIN:
+    timestepSetup(tid);
+    break;
+  case EXEC_SUBDOMAIN:
+    subdomainSetup(tid);
+    break;
+  case EXEC_NONLINEAR:
+    jacobianSetup(tid);
+    break;
+  case EXEC_LINEAR:
+    residualSetup(tid);
+    break;
+  default:
+    break;
+  }
+}
+
+
+template<typename T>
+void
+ExecuteMooseObjectStorage<T>::addObject(MooseSharedPointer<T> object, THREAD_ID tid/*=0*/)
+{
+  // Update list of all objects
+  MooseObjectStorage<T>::addObject(object, tid);
+
+  // Update the execute flag lists of objects
+  MooseSharedPointer<SetupInterface> ptr = MooseSharedNamespace::dynamic_pointer_cast<SetupInterface>(object);
+  if (ptr)
```

Comment:
There should probably be an error when this fails. I don't think we want MOOSE to silently fail to add objects to containers.